Lib/typing.py

```diff
@@ -2412,10 +2414,76 @@ def utf8(value: bytes) -> bytes: ...
       def utf8(value: str) -> bytes: ...
       def utf8(value):
           # implementation goes here
+
+    The overloads for a function can be retrieved at runtime using the
+    get_overloads() function.
     """
+    try:
+        key = _get_key_for_callable(func)
+    except TypeError:
+        # Not a normal function; ignore.
+        pass
+    else:
+        # We're inlining get_overloads() here to avoid computing the key twice.
+        existing = _overload_registry.setdefault(key, [])
+        if existing:
+            # If we are registering a variant with a lineno below or equal to that of the
+            # most recent existing variant, we're probably re-creating overloads for a
+            # function that already exists. In that case, we clear the existing variants
+            # to avoid leaking memory.
+            firstlineno = _get_firstlineno(func)
+            if firstlineno is not None:
+                existing_lineno = _get_firstlineno(existing[-1])
+                if existing_lineno is not None and firstlineno <= existing_lineno:
+                    existing.clear()
+
+        existing.append(func)
     return _overload_dummy
 
 
+def _get_firstlineno(func):
+    # staticmethod, classmethod
+    func = getattr(func, "__func__", func)
+    if not hasattr(func, '__code__'):
+        return None
+    return func.__code__.co_firstlineno
+
+
+# {key: [overload]}
+_overload_registry = {}
+
+
+def get_overloads(func):
+    """Return all defined overloads for *func* as a sequence."""
+    key = _get_key_for_callable(func)
+    return _overload_registry.get(key, [])
+
+
+def clear_overloads(func=None):
+    """Clear all overloads for the given function (or all functions)."""
+    if func is None:
+        _overload_registry.clear()
+    else:
+        key = _get_key_for_callable(func)
+        _overload_registry.pop(key, None)
+
+
+def _get_key_for_callable(func):
+    """Return a key for the given callable.
+
+    This is used as a key in the overload registry.
+
+    If no key can be created (because the object is not of a supported type), raise
+    AttributeError.
+    """
+    # classmethod and staticmethod
+    func = getattr(func, "__func__", func)
+    try:
+        return f"{func.__module__}.{func.__qualname__}"
```

Comment:
It looks like we might be able to shave off a few nanoseconds by using a tuple as the registry key, rather than using the tuple to create an f-string:

```python
In [1]: def get_key(func):
   ...:     return f'{func.__module__}.{func.__qualname__}'
   ...:

In [2]: def get_key2(func):
   ...:     return func.__module__, func.__qualname__
   ...:

In [3]: def f(): ...

In [4]: %timeit get_key(f)
155 ns ± 2.72 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [5]: %timeit get_key2(f)
107 ns ± 0.693 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```

```suggestion
        return func.__module__, func.__qualname__
```

Reply:
Good idea, I think that's also a little more correct in the edge case where a module within a package and a class within the package's `__init__.py` share the same name.